logstash-core/lib/logstash/plugins/event_factory_support.rb

```diff
@@ -0,0 +1,70 @@
+module LogStash
+  module Plugins
+    module EventFactorySupport
+
+      def event_factory
+        @event_factory ||= default_event_factory
+      end
+      attr_writer :event_factory
+
+      def default_event_factory
+        DefaultEventFactory::INSTANCE
+      end
+
+      def event_factory_builder
+        Builder.new self, default_event_factory
+      end
+
+      class Builder
+
+        def initialize(plugin, factory)
+          @plugin = plugin
+          @factory = factory
+        end
+
+        # @return an event factory
+        def build
```

Comment:
being a builder a named this `build` - but I also wanted the end result to set the new `@event_factory` for the plugin.
maybe it's nicer being explicit: `self.event_factory = event_factory_builder.with_target(target).build`
or having a `event_factory_builder.with_target(target).set_event_factory!` instead of `build`

Reply:
This feels subtly infectious and likely to be a source of surprises. I'm not sure we have found the right boundary.

A plugin developer should either be able to _know_ that they are working with a targeted event factory at its call-site, or they should not need to care about it _at all_. As implemented, we end up in a weird hybrid where the behaviour of `event_factory` is dependent on code paths that aren't immediately discoverable.

I think there are two directions we could go with this.

1.  Absorb `with_target` into `EventFactory`, making its use explicit all the way through

    Can we reasonably make `with_target` a part of the `EventFactory` API?

    Something like the following would give a lot of flexibility in _use_ while not being secretly magical. 

    ~~~ ruby
    class EventFactory ## base "abstract" class to be subclassed?
      def with_target(target_field_reference)
        return self if target_field_reference.nil?
      
        TargetedEventFactory.new(self, target_field_reference)
      end
    end
    ~~~

    ~~~ ruby
        def register
          @targeted_event_factory = event_factory.with_target(@target)
        end
        
        def decode(data)
          events = LogStash::Event.fromJson(data, @targeted_event_factory)
        end
    ~~~

2.  Make `EventFactorySupport#event_factory` _always_ automatically respect a provided `@target`. This feels even more magical and not _quite_ fully thought through, but perhaps closer to a useful and maintainable boundary.

    ~~~ ruby
    module EventFactorySupport
      def event_factory
        @event_factory ||= begin # TODO: threadsafety?
          ef = default_event_factory
          ef = TargetedEventFactory.new(ef, @target) unless @target.nil?
          ef
        end
      end
    end
    ~~~

Reply:
the way I was trying to approach this is as if we decided to later rely only on a Java `EventFactory` API, 

- having `EventFactory#withTarget` on the interface feels unfortunate, since the interface would be tight to a "targeted" factory implementation
- a `EventFactoryBuilder` of some kind might be necessary (later) if we decide on the explicit approach - instead of the `@event_factory` magically resolving itself based on `target` and `ecs_compatibility` fields. the explicitness should give us more control for whatever we decide to do later (at a cost of having slightly more code in the plugins)

My current candidate(s) would be to break the builder from magically setting the `event_factory` :
```ruby
  def register

    self.event_factory = event_factory_builder.with_target(target).event_factory
    
    # or being very explicit:
    
    self.event_factory = EventFactoryBuilder.new(default_event_factory).with_target(target).event_factory
    
  end
```